KAFKA-16518; Adding standalone argument for storage

muralibasani opened the pull request:
Resolves : https://issues.apache.org/jira/browse/KAFKA-16518

- Adds a new argument "standalone" to kafka-storage.sh
- If standalone mode, creates a checkpoint file in metadata dir ${kafkaConfig.metadataLogDir}/__cluster_metadata-0/


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

muralibasani commented:
@jsancio what do you think on this approach ?
it's abit not clear on creating the kRaftVersionRecord and votersRecord though.

```

./bin/kafka-storage.sh format --cluster-id oS80nAuvTOyliP131PWSFb --config ./config/kraft/server.properties --standalone
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/muralidharbasani/projects/others/kafka3/kafka/tools/build/dependant-libs-2.13.14/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/muralidharbasani/projects/others/kafka3/kafka/trogdor/build/dependant-libs-2.13.14/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Reload4jLoggerFactory]
metaPropertiesEnsemble=MetaPropertiesEnsemble(metadataLogDir=Optional.empty, dirs={/tmp/kraft-combined-logs: EMPTY})
logDir : /tmp/kraft-combined-logs
Formatting /tmp/kraft-combined-logs with metadata.version 3.7-IV4.
bootstrapDirectory : org.apache.kafka.metadata.bootstrap.BootstrapDirectory@1fb19a0
Snapshot written to /tmp/kraft-combined-logs/__cluster_metadata-0
muralidharbasani@Muralidhars-MacBook-Pro kafka % cat /tmp/kraft-combined-logs/__cluster_metadata-0/00000000000000000000-0000000000.checkpoint 
WYJ�� ������������������*
V�Ѣ�������������������$����"����?hY� �����������������%    
```

jsancio commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -452,21 +460,68 @@ object StorageTool extends Logging {
+
+    val builder = new RecordsSnapshotWriter.Builder()
+      .setKraftVersion(1)
+      .setVoterSet(Optional.empty())
```
If you create and set the `VoterSet` to contain this one replica, The builder will write the necessary `KRaftVersionRecord` and `VotersRecord`. To create a voter set you need the default listener name (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/raft/RaftManager.scala#L188), the replica id (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/raft/RaftManager.scala#L222), the directory id (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/SharedServer.scala#L260) and the advertise hostname and port (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaConfig.scala#L1237-L1243).

Looking at this code (https://github.com/apache/kafka/commit/e8d32563f323a245b3127394d97d6356e84f0294#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R2190-R2193), we are going to have to relax this to allowed advertise listeners as long as `controller.quorum.voters` is not set.

muralibasani replied:
@jsancio Thanks. when trying to create a voterset, VoterNode constructor is not accessible in core\tools, as VoterNode is part of raft package. Any suggestions ?

We need this public constructor https://github.com/apache/kafka/blob/8068a086a3f41b9f7e4d4a1dab3338f029089f23/raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java#L49

muralibasani replied:
trying this way 
```
    
  val voterMap: java.util.Map[Integer, VoterSet.VoterNode] = new util.HashMap()
      voterMap.put(1, new VoterSet.VoterNode(
        ReplicaKey.of(nodeId, Optional.of(directoryId)),
        Collections.singletonMap(controllerListenerNames, new InetSocketAddress(host, port)),
        new SupportedVersionRange(0.toShort, 1.toShort)
      ))
  val voterSet = new VoterSet(voterMap)

```

jsancio commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -452,21 +460,68 @@ object StorageTool extends Logging {
+      .setVoterSet(Optional.empty())
+      .setRawSnapshotWriter(rawSnapshotWriter).build(new MetadataRecordSerde)
+
+    // Append control records: KRaftVersionRecord and VotersRecord
+    appendControlRecords(builder)
```
This actually write those records as "metadata" records and not as controller records.

You don't need to call this method as the `RecordsSnapshotWriter.Builder.build(RecordSerde)` will write the necessary control record if you configure the builder correctly by calling `setKraftVersion` and `setVoterSet`.

jsancio reviewed (commented):
@muralibasani thanks for the feature. I left some comments.

muralibasani committed:
Adding test code

muralibasani committed:
Adding test code

sebastienviale committed:
MINOR: update all-latency-max, range-latency-avg|max and add prefix-scan documentation (#16182)

Reviewers: Matthias J. Sax <matthias@confluent.io>

JimGalasyn committed:
MINOR: Update docs for for KIP-671 (#16247)

Reviewers: Matthias J. Sax <matthias@confluent.io>

nicolasguyomar committed:
MINOR: Remove AK 1.0.0 reference from NetworkClient.java (#16223)

Reviewers: Justine Olshan <jolshan@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

ocadaruma committed:
KAFKA-16916: Disable blocking test temporarily (#16248)

The test runs forever. We disable the test temporarily to unblock CI

Reviewers: Luke Chen <showuon@gmail.com>

apoorvmittal10 committed:
KAFKA-16916: Fixing error in completing future (#16249)

Fix to complete Future which was stuck due the exception.getCause() throws an error.

The fix in the #16217 unblocked blocking thread but exception in catch block from blocking get call was wrapped in ExecutionException which is not the case when moved to async workflow hence getCause is not needed.

Reviewers: Luke Chen <showuon@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

soarez committed:
MINOR: Fix broken ReassignPartitionsCommandTest test (#16251)

KAFKA-16606 (#15834) introduced a change that broke
ReassignPartitionsCommandTest.testReassignmentCompletionDuringPartialUpgrade.

The point was to validate that the MetadataVersion supports JBOD
in KRaft when multiple log directories are configured.
We do that by checking the version used in
kafka-features.sh upgrade --metadata, and the version discovered
via a FeatureRecord for metadata.version in the cluster metadata.

There's no point in checking inter.broker.protocol.version in
KafkaConfig, since in KRaft, that configuration is deprecated
and ignored — always assuming the value of MINIMUM_KRAFT_VERSION.

The broken that was broken sets inter.broker.protocol.version in
KRaft mode and configures 3 directories. So alternatively, we
could change the test to not configure this property.
Since the property isn't forbidden in KRaft mode, just ignored,
and operators may forget to remove it, it seems better to remote
the fail condition in KafkaConfig.

Reviewers: Luke Chen <showuon@gmail.com>

soarez committed:
KAFKA-16886: Detect replica demotion in AssignmentsManager (#16232)

JBOD Brokers keep the Controller up to date with replica-to-directory
placement via AssignReplicasToDirsRequest. These requests are queued,
compacted and sent by AssignmentsManager.

The Controller returns the error NOT_LEADER_OR_FOLLOWER when handling
a AssignReplicasToDirsRequest from a broker that is not a replica.

A partition reassignment can take place, removing the Broker
as a replica before the AssignReplicasToDirsRequest successfully
reaches the Controller. AssignmentsManager retries failed
requests, and will continuously try to propagate this assignment,
until the Broker either shuts down, or is added back as a replica.

When encountering a NOT_LEADER_OR_FOLLOWER error, AssignmentsManager
should assume that the broker is no longer a replica, and stop
trying to propagate the directory assignment for that partition.

Reviewers: Luke Chen <showuon@gmail.com>

FrankYang0529 committed:
KAFKA-16878 Remove powermock and easymock from code base (#16236)

Reviewers: TaiJuWu <tjwu1217@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

gongxuanzhang committed:
KAFKA-10787 Apply spotless to transaction-coordinator and server-common (#16172)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

gaurav-narula committed:
KAFKA-16920: close BrokerLifecycleManager in tests (#16252)

Tests in BrokerLifecycleManagerTest do not close BrokerLifecycleManager
if an assertion fails.

This change makes BrokerLifecycleManager an instance variable that is
closed in an `@AfterEach` method.

Reviewers: Igor Soarez <i@soarez.me>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 18) Better assignment testing (#16201)

Added more testing for the StickyTaskAssignor, which includes a large-scale test with rack aware enabled.

Also added a no-op change to StreamsAssignmentScaleTest.java to allow for rack aware optimization testing.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 20) Custom task assignment configuration fix (#16245)

The StreamsConfig class was not parsing the new task assignment
configuration flag properly, which made it impossible to properly
configure a custom task assignor.

This PR fixes this and adds a bit of INFO logging to help users diagnose
assignor misconfiguration issues.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

chiacyu committed:
KAFKA-16885 Renamed the enableRemoteStorageSystem to isRemoteStorageSystemEnabled (#16256)

Reviewers: Kamal Chandraprakash <kchandraprakash@uber.com>, Chia-Ping Tsai <chia7712@gmail.com>

KrishVora01 committed:
modifying Readme for Docker official images (#16226)

This PR aims to modify the README file under /docker, to include the steps to release the Docker Official Images.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

KrishVora01 committed:
[MINOR] KAFKA-16373: KIP-1028: Modifying dockerfile comments (#16261)

This PR aims to add the replace the dockerfile comments under the jvm/dockerfile  Dockerfile, with updated url.
The original comment read
# Get kafka from https://archive.apache.org/dist/kafka and pass the url through build arguments . For DOI dockerfiles, we replace this with
# Get Kafka from https://downloads.apache.org/kafka, url passed as env var, for version {kafka_version} .

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

moscicky committed:
MINOR: Add fast_commit ext4 config to ops docs (#15496)


Reviewers: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: Piotr Rzysko <piotr.rzysko@allegro.com>

tinaselenge committed:
KAFKA-15718: Refactor UncleanLeaderElectionTest to enable KRaft later (#16157)

Refactor UncleanLeaderElectionTest to allow to enable KRaft later

Reviewers: Luke Chen <showuon@gmail.com>

loicgreffier committed:
KAFKA-16448: Add ProcessingExceptionHandler in Streams configuration (#16188)

This PR is part of KAFKA-16448 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR brings ProcessingExceptionHandler in Streams configuration.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: sebastienviale <sebastien.viale@michelin.com>

Reviewer: Bruno Cadonna <cadonna@apache.org>

riedelmax committed:
KAFKA-14509; [4/4] Handle includeAuthorizedOperations in ConsumerGroupDescribe API (#16158)

This patch implements the handling of `includeAuthorizedOperations` flag in the ConsumerGroupDescribe API.

Reviewers: David Jacot <djacot@confluent.io>

mimaison committed:
MINOR: Improve documentation of default.replication.factor config (#16243)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dajac committed:
MINOR: Rename uniform assignor's internal builders (#16233)

This patch renames the uniform assignor's builders to match the `SubscriptionType` which is used to determine which one is called. It removes the abstract class `AbstractUniformAssignmentBuilder` which is not necessary anymore. It also applies minor refactoring.

Reviewers: Ritika Reddy <rreddy@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

kamalcph committed:
KAFKA-15776: Support added to update remote.fetch.max.wait.ms dynamically (#16203)

Reviewers: Satish Duggana <satishd@apache.org>, Luke Chen <showuon@gmail.com>

appchemist committed:
KAFKA-16764: New consumer should throw InvalidTopicException on poll when invalid topic in metadata. (#16043)

Propagate metadata error from background thread to application thread.
So, this fix ensures that metadata errors are thrown to the user on consumer.poll()

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Philip Nee <pnee@confluent.io>, Lianet Magrans <lianetmr@gmail.com>

C0urante committed:
KAFKA-9228: Restart tasks on runtime-only connector config changes (#16053)

Reviewers: Greg Harris <greg.harris@aiven.io>

showuon committed:
add slf4jlog4j in tool (#16260)

In #12148 , we removed log4jAppender dependency, and add testImplementation dependency for slf4jlog4j lib. However, we need this runtime dependency in tools module to output logs. (ref) Adding this dependency back.

Note: The slf4jlog4j lib was added in log4j-appender dependency. Since it's removed, we need to explicitly declare it.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, Edoardo Comar <ecomar@uk.ibm.com>

ShivsundarR committed:
KAFKA-16724: Added support for fractional throughput and monotonic payload in kafka-producer-perf-test.sh

Added support for fractional throughput and monotonic payload in kafka-producer-perf-test.sh.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka#KIP932:QueuesforKafka-kafka-producer-perf-test.sh

Reviewers: Andrew Schofield <aschofield@confluent.io>, Manikumar Reddy <manikumar.reddy@gmail.com>

gongxuanzhang committed:
KAFKA-10787 Apply spotless to tools and tools-api module (#16262)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

adixitconfluent committed:
KAFKA-16914: Added share group dynamic and broker configs (#16268)

KIP-932 introduces a bunch of broker and dynamic configs for share groups. This PR adds those new configs. The changes include:

1. Defined ShareGroupConfigs class which stores various share group configurations.
2. Use the defined share configs in KafkaConfig.scala for making it available to BrokerServer
3. Adds a few tests to validate the conditions on these new configs.


 Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

KrishVora01 committed:
KAFKA-16373: KIP-1028: Adding 3.7.0 docker official images static assets (#16267)

This PR aims to add the static Dockerfile and scripts for AK 3.7.0 version. As mentioned in KIP-1028 this PR aims to start the release of the kafka:3.7.0 Docker Official image. This will also help us validate the process and allow us to address any changes suggested by Dockerhub before the 3.8.0 release.

The static Dockerfile and scripts have been generated via the github actions workflows and scripts added as part of https://github.com/apache/kafka/pull/16027. The reports of build and testing the 3.7.0 Docker official image are below.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

KrishVora01 committed:
KAFKA-16373: KIP-1028: Modfiying download url for kafka dockerfile (#16281)

This PR modifies the download url from https://downloads.apache.org/kafka/ to https://archive.apache.org/dist/kafka/ as the former is not permanent.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

kamalcph committed:
KAFKA-16904: Metric to measure the latency of remote read requests (#16209)

Reviewers: Satish Duggana <satishd@apache.org>, Christo Lolov <lolovc@amazon.com>, Luke Chen <showuon@gmail.com>

nizhikov committed:
KAFKA-16520: Support KIP-853 in DescribeQuorum (#16106)

Add support for KIP-953 KRaft Quorum reconfiguration in the DescribeQuorum request and response.
Also add support to AdminClient.describeQuorum, so that users will be able to find the current set of
quorum nodes, as well as their directories, via these RPCs.

Reviewers: Luke Chen <showuon@gmail.com>, Colin P. McCabe <cmccabe@apache.org>, Andrew Schofield <aschofield@confluent.io>

dajac committed:
KAFKA-16930; UniformHeterogeneousAssignmentBuilder throws NPE when one member has no subscriptions (#16283)

Fix the following NPE:

```
java.lang.NullPointerException: Cannot invoke "org.apache.kafka.coordinator.group.assignor.MemberAssignment.targetPartitions()" because the return value of "java.util.Map.get(Object)" is null
	at org.apache.kafka.coordinator.group.assignor.GeneralUniformAssignmentBuilder.canMemberParticipateInReassignment(GeneralUniformAssignmentBuilder.java:248)
	at org.apache.kafka.coordinator.group.assignor.GeneralUniformAssignmentBuilder.balance(GeneralUniformAssignmentBuilder.java:336)
	at org.apache.kafka.coordinator.group.assignor.GeneralUniformAssignmentBuilder.buildAssignment(GeneralUniformAssignmentBuilder.java:157)
	at org.apache.kafka.coordinator.group.assignor.UniformAssignor.assign(UniformAssignor.java:84)
	at org.apache.kafka.coordinator.group.consumer.TargetAssignmentBuilder.build(TargetAssignmentBuilder.java:302)
	at org.apache.kafka.coordinator.group.GroupMetadataManager.updateTargetAssignment(GroupMetadataManager.java:1913)
	at org.apache.kafka.coordinator.group.GroupMetadataManager.consumerGroupHeartbeat(GroupMetadataManager.java:1518)
	at org.apache.kafka.coordinator.group.GroupMetadataManager.consumerGroupHeartbeat(GroupMetadataManager.java:2254)
	at org.apache.kafka.coordinator.group.GroupCoordinatorShard.consumerGroupHeartbeat(GroupCoordinatorShard.java:308)
	at org.apache.kafka.coordinator.group.GroupCoordinatorService.lambda$consumerGroupHeartbeat$0(GroupCoordinatorService.java:298)
	at org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime$CoordinatorWriteEvent.lambda$run$0(CoordinatorRuntime.java:769)
	at org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime.withActiveContextOrThrow(CoordinatorRuntime.java:1582)
	at org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime.access$1400(CoordinatorRuntime.java:96)
	at org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime$CoordinatorWriteEvent.run(CoordinatorRuntime.java:767)
	at org.apache.kafka.coordinator.group.runtime.MultiThreadedEventProcessor$EventProcessorThread.handleEvents(MultiThreadedEventProcessor.java:144)
	at org.apache.kafka.coordinator.group.runtime.MultiThreadedEventProcessor$EventProcessorThread.run(MultiThreadedEventProcessor.java:176) 
```

Reviewers: Lianet Magrans <lianetmr@gmail.com>, Justine Olshan <jolshan@confluent.io>

C0urante committed:
MINOR: Fix flaky test ConnectWorkerIntegrationTest::testReconfigureConnectorWithFailingTaskConfigs (#16273)

Reviewers: Greg Harris <greg.harris@aiven.io>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 21) UUID to ProcessId migration (#16269)

This PR changes the assignment process to use ProcessId instead of UUID.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

C0urante committed:
MINOR: Wait for embedded clusters to start before using them in Connect OffsetsApiIntegrationTest (#16286)

Reviewers: Greg Harris <greg.harris@aiven.io>

abhijeetk88 committed:
KAFKA-15265: Integrate RLMQuotaManager for throttling copies to remote storage (#15820)

- Added the integration of the quota manager to throttle copy requests to the remote storage. Reference KIP-956
- Added unit-tests for the copy throttling logic.

Reviewers: Satish Duggana <satishd@apache.org>, Luke Chen <showuon@gmail.com>, Kamal Chandraprakash<kamal.chandraprakash@gmail.com>

muralibasani committed:
KAFKA-16922 Adding unit tests for NewTopic  (#16255)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 22) Add RackAwareOptimizationParams and other minor TaskAssignmentUtils changes (#16294)

We now provide a way to more easily customize the rack aware
optimizations that we provide by way of a configuration class called
RackAwareOptimizationParams.

We also simplified the APIs for the optimizeXYZ utility functions since
they were mutating the inputs anyway.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

cadonna committed:
KAFKA-10199: Enable state updater by default (#16107)

We have already enabled the state updater by default once.
However, we ran into issues that forced us to disable it again.
We think that we fixed those issues. So we want to enable the
state updater again by default.

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Matthias J. Sax <matthias@confluent.io>

dajac committed:
KAFKA-16770; [2/2] Coalesce records into bigger batches (#16215)

This patch is the continuation of https://github.com/apache/kafka/pull/15964. It introduces the records coalescing to the CoordinatorRuntime. It also introduces a new configuration `group.coordinator.append.linger.ms` which allows administrators to chose the linger time or disable it with zero. The new configuration defaults to 10ms.

Reviewers: Jeff Kim <jeff.kim@confluent.io>, Justine Olshan <jolshan@confluent.io>

wernerdv committed:
MINOR: Use predetermined dir IDs in ReplicationQuotasTest

Use predetermined directory IDs instead of Uuid.randomUuid() in ReplicationQuotasTest.

Reviewers: Igor Soarez <soarez@apple.com>

abhijeetk88 committed:
KAFKA-15265: Dynamic broker configs for remote fetch/copy quotas (#16078)

Reviewers: Kamal Chandraprakash<kamal.chandraprakash@gmail.com>, Satish Duggana <satishd@apache.org>

gongxuanzhang committed:
KAFKA-10787 Apply spotless to connect module (#16299)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

gongxuanzhang committed:
KAFKA-10787 Apply spotless to log4j-appender, trogdor, jmh-benchmarks, examples, shell and generator (#16296)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

tinaselenge committed:
KAFKA-16865: Add IncludeTopicAuthorizedOperations option for DescribeTopicPartitionsRequest (#16136)


Reviewers: Mickael Maison <mickael.maison@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>, Calvin Liu <caliu@confluent.io>, Andrew Schofield <andrew_schofield@live.com>, Apoorv Mittal <amittal@confluent.io>

edoardocomar committed:
KAFKA-16570 FenceProducers API returns "unexpected error" when succes… (#16229)

KAFKA-16570 FenceProducers API returns "unexpected error" when successful

* Client handling of ConcurrentTransactionsException as retriable
* Unit test
* Integration test

Reviewers: Chris Egerton <chrise@aiven.io>, Justine Olshan <jolshan@confluent.io>

ivanyu committed:
KAFKA-8206: KIP-899: Allow client to rebootstrap  (#13277)

This commit implements KIP-899: Allow producer and consumer clients to rebootstrap. It introduces the new setting `metadata.recovery.strategy`, applicable to all the types of clients.

Reviewers: Greg Harris <gharris1727@gmail.com>, Rajini Sivaram <rajinisivaram@googlemail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 23) More TaskAssignmentUtils tests (#16292)

Also moved the assignment validation test from StreamsPartitionAssignorTest to TaskAssignmentUtilsTest.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

m1a2st committed:
KAFKA-16897 Move OffsetIndexTest and OffsetMapTest to storage module (#16244)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

kamalcph committed:
KAFKA-16890: Compute valid log-start-offset when deleting overlapping remote segments (#16237)

The listRemoteLogSegments returns the metadata list sorted by the start-offset. However, the returned metadata list contains all the uploaded segment information including the duplicate and overlapping remote-log-segments. The reason for duplicate/overlapping remote-log-segments cases is explained [here](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogLeaderEpochState.java#L103).

The list returned by the RLMM#listRemoteLogSegments can contain the duplicate segment metadata at the end of the list. So, while computing the next log-start-offset we should take the maximum of segments (end-offset + 1).

Reviewers: Satish Duggana <satishd@apache.org>

JimGalasyn committed:
MINOR: Remove Java 7 example code (#16308)

Reviewers: Matthias J. Sax <matthias@confluent.io>

C0urante committed:
KAFKA-16935: Automatically wait for cluster startup in embedded Connect integration tests (#16288)

Reviewers: Greg Harris <gharris1727@gmail.com>

tinaselenge committed:
MINOR: Clean up for KafkaAdminClientTest (#16285)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-16938 non-dynamic props gets corrupted due to circular reference between DynamicBrokerConfig and DynamicConfig. (#16302)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

C0urante committed:
MINOR: Add readiness check for connector and separate Kafka cluster in ExactlyOnceSourceIntegrationTest::testSeparateOffsetsTopic (#16306)

Reviewers: Greg Harris <gharris1727@gmail.com>

lianetm committed:
MINOR: Improving log for outstanding requests on close and cleanup (#16304)

Reviewers: Andrew Schofield <aschofield@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

frankvicky committed:
KAFKA-16901 Add unit tests for ConsumerRecords#records(String) (#16227)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

brenden20 committed:
KAFKA-16557 Implemented OffsetFetchRequestState toStringBase and added a test for it (#16291)

Reviewers: Lianet Magrans <lianetmr@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 24) internal TaskAssignor rename to LegacyTaskAssignor (#16318)

Since the new public API for TaskAssignor shared a name, this rename will prevent users from confusing the internal definition with the public one.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

gongxuanzhang committed:
KAFKA-16643 Add ModifierOrder checkstyle rule (#15890)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

m1a2st committed:
KAFKA-16921 [1/N] Migrate all junit 4 code to junit 5 for connect module (#16253)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dongnuo123 committed:
MINOR: Make online downgrade failure logs less noisy and update the timeouts scheduled in `convertToConsumerGroup` (#16290)

This patch: 
- changes the order of the checks in `validateOnlineDowngrade`, so that only when the last member using the consumer protocol leave and the group still has classic member(s), `online downgrade is disabled` is logged if the policy doesn't allow downgrade.
- changes the session timeout in `convertToConsumerGroup` from `consumerGroupSessionTimeoutMs` to `member.classicProtocolSessionTimeout().get()`.

Reviewers: David Jacot <djacot@confluent.io>

muralibasani committed:
Add standalone argument

muralibasani committed:
Merge branch 'kafka-16518' of github.com:muralibasani/kafka into kafka-16518
